### ISSUE TITLE
Cloudwatch RUM support

### ISSUE THREAD

**Issue by Matt-Butler** [aws/aws-cdk#17939]
### Description

https://aws.amazon.com/blogs/aws/cloudwatch-rum/

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rum-appmonitor.html

### Use Case

Needs CFN support: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/1001

### Proposed Solution

New Construct

### Other information

https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/1001

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by robertd** [comment]
https://github.com/aws/aws-cdk/pull/17844 added this construct. 

It has been released in `1.135.0` but not in `2.1.0` yet. 

This issue can be closed.

**Comment by Matt-Butler** [comment]
https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/aws-rum

**Comment by blimmer** [comment]
Awesome! I'm glad to see this is included as a `Cfn` class in newer versions. I'm on 2.x, so will keep an eye on it.

Moving forward, will it be common for new releases like this to make it into the 1.x branch, but not 2.x? Or is the difference between the cfn-spec between the 1.x and 2.x branches a short-term thing we'll only experience during the transition?

**Comment by robertd** [comment]
@blimmer https://github.com/aws/aws-cdk-rfcs/blob/master/text/0079-cdk-2.0.md#aws-cdk-maintenance-policy

![image](https://user-images.githubusercontent.com/31543/146092462-1b2c9a7f-ac21-450e-8278-c8df39749a07.png)

tl;dr - v1 will be fully supported (as it is today) for 6 months. We'll probably see features getting ported back and forth between the versions until 2.x becomes main.

**Comment by blimmer** [comment]
Awesome - thanks for the information. I opened https://github.com/aws/aws-cdk/issues/18024 with some ideas on how we could expose this type of information more readily to aws-cdk users. I'd love to hear your feedback if you have any!

**Comment by blimmer** [comment]
@robertd Awesome - thanks for the information. I opened https://github.com/aws/aws-cdk/issues/18024 with some ideas on how we could expose this type of information more readily to aws-cdk users. I'd love to hear your feedback if you have any!

**Comment by ckifer** [comment]
Is there any way to compose the JS snippet to inject into your website from the attributes on this Cfn construct? I don't think so after attempting but just want to confirm.

**Comment by blimmer** [comment]
If anyone's looking for some help using the Cfn classes, here's the code I used to get this working:

```ts
import * as cdk from 'aws-cdk-lib';
import { aws_cognito as cognito, aws_iam as iam, aws_rum as rum } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class MonitoringStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const siteDomainName = 'example.com'; // TODO: change this to your site!

    // The identity pool and role are usually created automatically by the UI-based wizard.
    // I generated one via the wizard to determine the proper settings.
    // See also: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-RUM-get-started-authorization.html
    const identityPool = new cognito.CfnIdentityPool(this, 'RumAppIdentityPool', {
      identityPoolName: `${siteDomainName}-rum-monitor`,
      allowUnauthenticatedIdentities: true,
    });

    const unauthenticatedRumRole = new iam.Role(this, 'UnauthenticatedRumRole', {
      roleName: `${siteDomainName}-RUM-UnauthenticatedRole`,
      assumedBy: new iam.FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'unauthenticated' },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
      inlinePolicies: {
        RUMPutBatchMetrics: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              actions: ['rum:PutRumEvents'],
              resources: [this.formatArn({ service: 'rum', resource: 'appmonitor', resourceName: siteDomainName })],
            }),
          ],
        }),
      },
    });

    new cognito.CfnIdentityPoolRoleAttachment(this, 'RumAppRoleAttachment', {
      identityPoolId: identityPool.ref,
      roles: {
        unauthenticated: unauthenticatedRumRole.roleArn,
      },
    });

    new rum.CfnAppMonitor(this, 'RumAppMonitor', {
      name: siteDomainName,
      cwLogEnabled: false, // stores telemetry for > 30 days in CW - disable for now
      domain: siteDomainName,
      appMonitorConfiguration: {
        allowCookies: true,
        enableXRay: false,
        sessionSampleRate: 1, // sample all sessions
        telemetries: ['errors', 'performance', 'http'],
        identityPoolId: identityPool.ref,
        guestRoleArn: unauthenticatedRumRole.roleArn,
      },
    });
  }
}

```

**Comment by madeline-k** [comment]
Thanks @blimmer for providing an example! Closing this issue as aws-rum is support in 1.x and 2.x now.

**Comment by sebsonjura** [comment]
@blimmer: It is your 'fault' that i will be 'home' in time today. Thanks for taking the time.

**Comment by simonireilly** [comment]
Added a Cloudformation custom resource that injects the rum script into the target bucket for the s3 origin: https://github.com/simonireilly/canary-stack/blob/main/lib/rum/rum-constrcut.ts

Also includes a full example of how to add some canaries and associate them via tags :+1:

```ts
/**
 * The RUM custom resource can be used to setup Real User Monitoring using AWS
 *
 * The resource itself creates all the required infrastructure.
 *
 * A Cloudformation custom resource uploads the rum script to the s3 bucket that
 * the website is deployed to:
 *
 * @example
 * const rum = new Rum(this, "SiteRum", {
 *   topLevelDomain: "*.s3-website-eu-west-1.amazonaws.com",
 *   appMonitorName: "canary-stack-rum",
 *   s3Bucket: websiteBucket,
 * });
 *
 */
 ```

**Comment by simonireilly** [comment]
> @simonireilly dropping an updated link to your construct because that link is broken :) https://github.com/simonireilly/canary-stack/blob/main/lib/rum/rum-construct.ts

Thanks, I've updated 👍
